### ISSUE TITLE
bug: attach 后输入框残影且 ←/ctrl+] 无法 detach —— jiggle frameStart 快速通道把 hold 提前还原，净零塌陷后零自愈（#25 回归窗口）

### ISSUE THREAD

**Issue by zhuxixi** [zhuxixi/pi-agent-board#42]
## 现象

attach 一个已存在的 session 后，**有概率**进来看到子 session 的输入框区域渲染损坏：编辑器框体出现新旧两帧错位叠加——边框线重复、下边框线上嵌入 `◊◊` 等杂散字符、框体下方多出残留行片段（本地复现于 2026-08-29，截图存档在 /tmp/pi-clipboard-f976c121-3eb6-470c-aa44-c6f5593afe29.png，可拖入本 issue）。

更影响使用的是**交互被困死**：此时按 `←` 无法 detach，按 `ctrl+]` 也同样无效。唯一的恢复路径是用户摸索出来的：

1. 按 `↑` —— 子 session 的 pi 编辑器召回上一条历史输入，编辑器区域重绘；
2. 按 `↓` —— 清掉召回的内容，编辑器再次重绘为干净的空输入行；
3. 此时按 `←` 才能 detach 回 dashboard。

即：一次按键迫使子 session 重绘编辑器区域后，attach 表面的 buffer 被"治愈"，detach 才恢复可用。

## 根因定位

这是两个 bug 叠加：**jiggle 自愈协议存在净零塌陷竞态**（主因）+ **detach 门禁把 ctrl+] 也绑死在不可靠的 buffer 启发式上**（次因，放大了主因的影响）。

### 主因：shrink-and-hold 的 frameStart 快速通道会把 hold 提前取消，重新引入 #25 要消灭的净零塌陷

attach 流程依赖"重放屏幕日志重建近似画面 → jiggle 逼迫子 session fullRender 把重建误差全部覆盖"。回放重建本身可以是错的（见下），设计上的兜底是 shrink-and-hold（#25）：缩 1 列 1 行并**保持**，直到观察到子 session 吐出 `\x1b[2J` 全屏清除才还原尺寸——没有 clear 就没有还原，净零塌陷不可能发生。

但 `feed()` 里的 frameStart 快速通道破坏了这个不变量（`src/core/pty-attach-jiggle-controller.mjs:174-186`）：

```js
if (result.frameStartFound && !tuiFrameSeen) {
    tuiFrameSeen = true;
    clearG1Timer();
    if (held) {
        restoreIfHeld(); // fast path: 一看到任何 TUI 帧就立刻还原尺寸
    }
    ...
}
```

这个通道是为**冷启动**设计的（issue #25：子 TUI 在我们 shrink 之后才开始渲染，首帧以 shrunk 尺寸建基线、永远不发 clear，所以看到首帧就还原尺寸制造宽度差）。但对**热 session**（attach 时子 pi 正在渲染/有输出在途），它是一个一碰就发的扳机：

1. connect 后 `start()` 发出 `resize(cols,rows)` + `resize(cols-1,rows-1)`，hold 住；
2. 子 session 在 shrink 生效**之前**发出的差分帧（流式输出的尾巴、状态 tick 等）此时才经 socket 送达——帧以 `\x1b[?2026h` 开头但**不含** `\x1b[2J`；
3. `feed()` 看到 frameStart → 快速通道立刻 `restoreIfHeld()` 把尺寸改回去；
4. 还原 resize 与 shrink 在子进程侧合并：pi-tui 的 resize 处理只是非强制的 `requestRender()`（pi-tui `dist/tui.js:433`），且有 16ms 渲染节流（`MIN_RENDER_INTERVAL_MS`，`tui.js:123/532`）。如果还原落在子进程下一次 doRender 之前，子进程查询到的尺寸与基线**净零不变** → `widthChanged=false` → 只走差分渲染，内容没变时甚至**一个字节都不输出**；
5. 此后：`tuiFrameSeen=true` 使 G1 定时器被清除；退避重试链在 hold 协议下是"纯倒计时"（只数预算、从不重发 resize），预算耗尽后的 G2 `restoreIfHeld()` 也是 no-op（已还原）。**再没有任何机制去戳子 session 重绘**——buffer 永远停留在回放重建的近似画面上。

于是用户看到的就是未被治愈的重建残影。触发概率取决于 attach 瞬间子 session 是否有在途输出：session 刚跑完流式输出、dashboard 状态刷新后立刻 attach 的场景最容易踩中——与"有的时候"的表现一致。

### 帮凶：回放重建本身不保证正确，`\n` 对齐启发式对 pi-tui 帧无效

`replayScreenLog()`（`src/ui/pty-attach.ts:899-920`）从日志尾部 60KB 处硬切，仅"跳到第一个 `\n` 之后"来避免注入半个转义序列。但 pi-tui 的帧内部本来就充满 `\r\n`（fullRender 行间分隔符，差分帧也用相对光标移动 `\x1b[NB/A` + `\r`），所以切点几乎总是落在**某一帧的中间**：

- 若切在 fullRender 中间：后续行从切点处顺序向下写，起始行错位；
- 若切在差分帧中间：相对光标移动基于错误的起点，行会写到错误的绝对行号上。

编辑器框体错位叠加、边框线上嵌入杂散字符，正是相对寻址帧错位的典型痕迹。回放不准本来是可接受的（jiggle 会兜底全量覆盖），但主因让兜底失效后，回放误差就直接暴露给用户了。

### 次因：detach 门禁把 ctrl+] 也绑死在可能损坏的 buffer 上

`handleInput`（`src/ui/pty-attach.ts:232-246`）中 `←` 和 `ctrl+]` 共用同一个门禁：

```js
if (DETACH_KEYS.has(data) || matchesKey(data, Key.left) || matchesKey(data, Key.ctrl("]"))) {
    if (this.childInputLooksEmpty()) { this.detach(); return; }
    // 否则按键被转发给子 session —— 用户被困住
}
```

`childInputLooksEmpty()`（`pty-attach.ts:308-314`）读 headless xterm 光标所在行，交给 `isProbablyEmptyPiInputLine()`（`src/core/pty-input.mjs:11`）判断。当 buffer 因主因而损坏时，光标行上是边框残片/`◊◊` 等字符 → 判定为"子编辑器有内容" → `←` **和** `ctrl+]` 都被转发给子 session，用户没有任何逃生键。

`←` 需要这个门禁是合理的（它也是子编辑器的左移光标键）；但 `ctrl+]` 不是任何编辑键，绑在同一个启发式上导致显式逃生通道一并失效。

## 修复方向建议（按优先级）

1. **堵住净零塌陷**（主因）：快速通道的还原应该以"clear 之后的帧"为准，或者还原后若一定窗口内没观察到 `\x1b[2J` 就重新 shrink（re-arm 一次新的 hold）；更保守的做法是让退避链在 hold 期间真正的周期性重发 shrink 脉冲而不是纯倒计时。核心不变量：*只要没见到 clear，就必须保证子进程最终能观察到一个非零的尺寸差。*
2. **ctrl+] 绕过输入框门禁**：`DETACH_KEYS`/`ctrl+]` 不经 `childInputLooksEmpty()` 直接 detach（`←` 保留门禁）。这样即使主因复发，用户也永远有确定性的逃生键，不用再靠 ↑↓ 摸黑自愈。
3. **回放切点对齐到帧边界**（降低主因复发时的视觉损伤）：从尾部窗口里最后一个 `\x1b[?2026h`（帧起点）或最后一个 `\x1b[2J`（全量帧起点）开始回放，而不是按 `\n` 切。
4. 可观测性：jiggle 链预算耗尽/快速通道触发且无 clear 跟进时，写一条 diagnostics 日志，便于事后确认命中率。

## 关联

- #25（shrink-and-hold 协议本身——本 issue 是其快速通道引入的回归窗口）
- #10（jiggle 链断后零自愈的同类症状）
- #11（attach 加固：检测光标失同步并触发补救 jiggle——落地后可作为本 bug 的兜底探测）
- #2（jiggle 自愈的原始 issue）

## 环境

pi-agent-board v0.4.3（23c7c46），Linux，attach 目标为刚结束输出的 idle session。

**Comment by zhuxixi** [comment]
## 复现与修复验证（已在本地完成双向确认）

### 复现：两层确定性复现，修复前均稳定失败

**单元层**（控制器决策缺陷）：`start()` 后喂一帧只含 `\x1b[?2026h` 不含 `\x1b[2J` 的在途差分帧 → 快速通道立刻还原 → 排干全部退避计时器 → 断言"必须再发一次 shrink 重戳"。**修复前失败**：`shrinks=1`，即净零塌陷后控制器永远不再戳子进程，buffer 永远停留在回放残影上。

**端到端层**（真 runner + 真 PTY + 真 socket 协议 + 忠实模拟 pi-tui 行为的 hot stub 子进程，`test-support/fake-hot-tui-pi.mjs`：resize 处理带延迟合并、读当前尺寸——净零即沉默）：子进程已热、每 15ms 发差分帧时 attach → 快速通道被在途帧误触 → 还原落在子进程合并窗口内 → 沉默。**修复前失败**：8s 内等不到 `\x1b[2J`（永不自愈）。**修复后通过**：~1.2s 内完成 heal 且 PTY 回到原始尺寸，连跑 3 次稳定。

### 修复方案（未提交，工作区待审）

1. **`src/core/pty-attach-jiggle-controller.mjs`**（主因，+52/-7）：新增 G6 post-restore verify——frameStart 快速通道还原后重启退避链，900ms 验证窗口内没等到 clear 则由链上的 `ensureHold()` 重新 shrink（单独落地，不存在可与之合并的 restore，必然让子进程看到真实宽度差）；预算仍由原 backoff 兜底（G2），无副作用期行为变化。shell/冷启动路径因 `!tuiFrameSeen` 守卫保持逐字节不变。
2. **`src/ui/pty-attach.ts`**（逃生通道，+14/-5）：`ctrl+]` 不再经 `childInputLooksEmpty()` 门禁，无条件 detach；`←` 保留门禁。

### 对 QA 基线的影响确认

- 基线：CI = typecheck + 320 测试 + 覆盖率门禁（lines 85/funcs 80/branches 70）+ pack:dry，Node 22/24 双版本。
- 修复后本地 `npm run verify` 全绿：**325 测试全过**（320 → 325：新增 3 条控制器回归 + 1 条热 session e2e + 1 条 detach 门禁组件测试），覆盖率 91.9% lines / 74.7% branches / 89.5% funcs，均在门禁之上。
- **唯一被改动的既有测试是 2 条单元测试**（`G2: backoff budget exhausted without clear…` 和 `full G1 window passes…`）——它们此前把"快速通道还原后永不重戳"这一 bug 行为断言为正确契约，已更新为新契约并补充注释说明。其余 318 条测试（含 #25 冷启动 3 条 e2e、G1-G5 全部守卫）逐条未动、全部通过。
- 真 pi-tui 行为假设（resize → 非强制 requestRender → 净零静默）已从仓库 `node_modules/@earendil-works/pi-tui/dist/tui.js:433` 的源码直接核实。

### 遗留（建议另开 issue 或在此跟进）

- 回放切点对齐到帧边界（最后一个 `\x1b[?2026h` / `\x1b[2J`）而非按 `\n` 硬切——降低主因万一再复发时的视觉损伤，非本次修复必需。

**Comment by ccccyk0919** [comment]
第三方复现确认（Windows 环境）。补充观察供修复参考：

**环境**：Windows 11 · pi 0.84.4 · agent-board 0.4.3（本地 fork，已含已合并的 #44/#45 Windows 修复，不含本 issue 修复）

**现象**：attach 已有 session 后 `←` 无法 detach。恢复路径是：在输入框输入内容再清空后，`←` 恢复可用——与 #42 描述的 ↑/↓ 历史召回是同一恢复机制（按键迫使子 session 重绘编辑器区域，attach 表面 buffer 被"治愈"）。

**screen.log 观察**（attach 表面录制的渲染流）：
- attach 时回放重建，随后日志中出现多次 `\x1b[2J`（jiggle heal 痕迹），说明 shrink-and-hold 的 heal 路径在本环境曾生效；
- 但 heal 之后 `←` 依然失败过——即 detach 门禁对 buffer 光标行的启发式检测在屏幕"看起来正常"（无明显残影）的状态下也会误判：`childInputLooksEmpty()` 取到的光标行若落在 footer/边框等非输入行，`isProbablyEmptyPiInputLine` 返回 false，`←` 被转发给子进程而不是 detach。

**对修复方案的反馈**：`ctrl+]` 无条件 detach 的逃生通道方向正确。建议合并后评估 `←` 是否同样绕过门禁，或后续把门禁替换为可靠检测（例如通过 PTY host 控制 socket 查询子进程编辑器状态）——基于渲染内容的启发式检测在活渲染状态下也脆弱，可能成为后续同类 issue 的温床。

合并后我会同步上游并在本地验证。

**Comment by ccccyk0919** [comment]
补充一个新观察：**pi 思考（streaming）时 `←` 也无法回退**，即使输入框为空。

**现象**：attach 表面中，子 session 的 pi 正在思考/输出（Working 指示动画运行中），输入框为空，按 `←` 无反应（不回退、也不转发出可见效果）；思考结束后 `←` 恢复。

**screen.log 证据**（同一 attach 表面的渲染流）：
- 思考期间输入行（第 39 行）持续渲染为**空白 + 反白假光标**（`\x1b[7m \x1b[27m` + 清行，日志中出现 245 次）——输入框确认是空的，门禁的"编辑中不抢键"保护并不应该触发；
- 但 Working 指示行（第 36 行）每帧渲染为 `⠙ Working...`（**有文字**）；
- 即：streaming 期间 xterm 的光标行并不在空输入行上（每帧重绘 Working 行 + 输出区滚动都会移动它），`childInputLooksEmpty()` 取到 Working/消息等非空行 → `isProbablyEmptyPiInputLine` 返回 false → `←` 被转发给子进程而不是 detach。

**与 #42 修复的关系**：这是同一个门禁（`childInputLooksEmpty` 的 buffer 光标行启发式）的另一个稳定触发场景——不是残影塌陷，而是 streaming 期间光标行必然离开输入行。当前修复方案让 `ctrl+]` 绕门禁可以缓解，但 `←` 保留门禁意味着**此场景在修复合并后依然存在**。

建议：`←` 门禁同样考虑绕过（detach 不打断子进程，思考中回到 dashboard 应被允许），或把门禁改为不依赖光标行的检测（如 PTY host 通过控制 socket 查询子进程编辑器状态）。

**Comment by ccccyk0919** [comment]
再补充实测确认：**思考（streaming）期间，"输入字符再删掉"的 workaround 也无效**——输入框清空后按 `←` 依然无法回退（非思考时该 workaround 有效，思考时无效）。

**机制细化**（screen.log 统计支撑）：

- 整个日志共 **1831 帧 Working 动画重绘**（`⠙⠹⠸…` 循环，仅重绘指示行第 36 行）；最近 60KB 内仍有 75 帧；
- 同一 60KB 内**编辑器行的假光标渲染出现 0 次**——streaming 期间差分渲染只画变更行（Working 行 + 消息区），编辑器区域不重绘、假光标不输出、光标定位序列不执行；
- 因此 streaming 期间 xterm 光标行**持续**停在 Working 指示行（内容 `⠙ Working...`，必判非空），`←` 门禁处于**持续失败状态**，不是瞬态；
- 输入字符会触发编辑器重绘把光标短暂拉回输入行，但删除后下一个动画帧（~100-120ms）又把光标拉回 Working 行——"输入再删掉"的窗口不足一个动画帧间隔，实际不可用。

**结论**：streaming 期间 `←` 门禁必然失败，这是用户最常遇到的场景（pi 思考时想回 dashboard 看别的会话）。#42 当前修复方案保留 `←` 门禁，意味着此场景在修复合并后**依然存在**。强烈建议 `←` 一并绕过门禁（或改为不依赖光标行的检测），否则建议至少把 `←` 的失败提示出来（如状态栏提示"输入为空但无法 detach"）以便用户知道用 `ctrl+]`。

**Comment by zhuxixi** [comment]
修复已提交并开 PR：#59（含 G6 根因修复、ctrl+] 透传决策落地、graceful detach、socket identity guard、child-aware runner shutdown；409/409 verify 全绿，4 轮独立 review 最后一轮 Ready to merge: Yes）。合并后本 issue 自动关闭。
